File: ModuleProjects/ParkingArea/UsingJAVA/src/main/java/com/kururu/simple/project/function/ParkingCar.java
package com.kururu.simple.project.function;

import static com.kururu.simple.project.constant.ParkingAreaConstants.WARN_MESSAGE_INPUT_ERROR;

import com.kururu.simple.project.constant.ParkingAreaEnums.END_BUSINESS_FLG;
import com.kururu.simple.project.constant.ParkingAreaEnums.CAR_SIZE;
import com.kururu.simple.project.dto.ParkingCarDto;
import com.kururu.simple.project.entity.EntryBook;
import com.kururu.simple.project.entity.EntryBookIdentity;
import com.kururu.simple.project.repository.condition.CountExistAreaCondition;
import com.kururu.simple.project.repository.EntryBookRepository;
import com.kururu.simple.project.utility.common.DateComponent;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.Objects;

/**
 * <h2>Parking Area [Park new car to free area]</h2>
 *
 * @author Rick00Kim dreamx119@gmail.com
 */
@Component
@Slf4j
public class ParkingCar extends AbstractFunction {

    private static final String PARKING_CAR_DTO = "parkingCarDto";

    @Autowired
    private EntryBookRepository entryBookRepository;

    @Autowired
    private DateComponent dateComponent;

    @Override
    protected RESULT_STATUS input() {

        final ParkingCarDto parkingCarDto = ParkingCarDto.builder().build();
        try {
            parkingCarDto.setVehicleNumber(
                    userInputComponent.getUserInput("Vehicle number : "));
            parkingCarDto.setInputCarSize(
                    Objects.requireNonNull(
                            CAR_SIZE.getValue(
                                    userInputComponent.getUserInput("Car size : ")),
                            String.format(WARN_MESSAGE_INPUT_ERROR, "CAR_SIZE")));
        } catch (Exception e) {
            log.warn(String.format(WARN_MESSAGE_INPUT_ERROR, "INPUT"), e);
            return RESULT_STATUS.FAILURE;
        }

        functionMap.put(PARKING_CAR_DTO, parkingCarDto);

        return RESULT_STATUS.SUCCESS;
    }

    @Override
    protected RESULT_STATUS validate() {

        final ParkingCarDto parkingCarDto = (ParkingCarDto) functionMap.get(PARKING_CAR_DTO);
        final CountExistAreaCondition countExistAreaCondition = CountExistAreaCondition.builder()
                .lotNumber(currentLotInformation.getLotNumber())
                .carSize(parkingCarDto.getInputCarSize())
                .pairTermADay(dateComponent.getDateTermADay())
                .build();

        final int count = entryBookRepository.countExistEmptyArea(countExistAreaCondition);
        switch (parkingCarDto.getInputCarSize()) {
            case SMALL:
                if (count >= currentLotInformation.getAcceptableSmall())
                    return RESULT_STATUS.FAILURE;
            case MEDIUM:
                if (count >= currentLotInformation.getAcceptableMedium())
                    return RESULT_STATUS.FAILURE;
            case HEAVY:
                if (count >= currentLotInformation.getAcceptableHeavy())
                    return RESULT_STATUS.FAILURE;
        }

        return RESULT_STATUS.SUCCESS;
    }

    @Override
    protected RESULT_STATUS process() {

        final ParkingCarDto parkingCarDto = (ParkingCarDto) functionMap.get(PARKING_CAR_DTO);
        entryBookRepository.save(EntryBook.builder()
                .key(EntryBookIdentity.builder()
                        .vehicleNumber(parkingCarDto.getVehicleNumber())
                        .clientNumber("Non-Member")
                        .lotNumber(currentLotInformation.getLotNumber())
                        .build())
                .carSize(parkingCarDto.getInputCarSize())
                .arrivalTime(dateComponent.getCurrentTimestamp())
                .endBusinessFlg(END_BUSINESS_FLG.BUSINESS_NOT_ENDED)
                .build());

        return RESULT_STATUS.SUCCESS;
    }

}

File: ModuleProjects/ParkingArea/UsingPython/parking_area/functions/parking_car.py
from parking_area.dto.entry_book import EntryBook
from parking_area.functions.base_function import BaseFunction
from parking_area.constants.parking_area_constants import ParkingAreaConstants
from parking_area.constants.parking_area_enums import ParkingAreaEnums
from parking_area.database.entry_book_repo import EntryBookRepository
from parking_area.database.condition.check_empty_condition import CountExistEmptyAreaCondition
from parking_area.utilities.time_utility import get_start_on_today, get_end_on_today, get_now_datetime


class ParkingCar(BaseFunction):
    """Parking Area [Park new car to free area]

    """

    # Input field
    vehicle_number: str
    car_size: str
    function_name = "Parking Car"

    def input_func(self):
        # Get input variables
        self.vehicle_number = input("Vehicle number : ")
        input_car_size = input("Car size : ")

        # Check input variables(Return Failure if variables is null)
        # VEHICLE NUMBER
        if not self.vehicle_number or not self.vehicle_number.strip():
            print("INPUT Vehicle number is NULL")
            return ParkingAreaEnums.ResultStatusEnums.FAILURE
        # CAR SIZE
        if not input_car_size or not input_car_size.strip():
            print("INPUT Car size is Null")
            return ParkingAreaEnums.ResultStatusEnums.FAILURE
        else:
            if input_car_size not in ParkingAreaConstants.car_size_dict.keys():
                print("Car Size is invalid")
                return ParkingAreaEnums.ResultStatusEnums.FAILURE
            else:
                self.car_size = input_car_size
        # When all pass checking input variables, Return Success
        return ParkingAreaEnums.ResultStatusEnums.SUCCESS

    def validate_func(self):
        # Check Available parking area
        using_count = EntryBookRepository().count_exist_empty_area(
            CountExistEmptyAreaCondition(
                ParkingAreaConstants.CURRENT_LOT_INFORMATION.lot_number,
                self.car_size,
                get_start_on_today(),
                get_end_on_today()
            )
        )
        # Check available park area
        if self.car_size is '0':
            if using_count >= ParkingAreaConstants.CURRENT_LOT_INFORMATION.acceptable_small:
                return ParkingAreaEnums.ResultStatusEnums.FAILURE
        elif self.car_size is '1':
            if using_count >= ParkingAreaConstants.CURRENT_LOT_INFORMATION.acceptable_medium:
                return ParkingAreaEnums.ResultStatusEnums.FAILURE
        elif self.car_size is '2':
            if using_count >= ParkingAreaConstants.CURRENT_LOT_INFORMATION.acceptable_heavy:
                return ParkingAreaEnums.ResultStatusEnums.FAILURE
        # When all pass validate, Return Success
        return ParkingAreaEnums.ResultStatusEnums.SUCCESS

    def process_func(self):
        # Insert new Entry book
        EntryBookRepository().insert_one_entity(
            EntryBook(
                self.vehicle_number,
                'Non-Member',
                ParkingAreaConstants.CURRENT_LOT_INFORMATION.lot_number,
                self.car_size,
                get_now_datetime(),
                None,
                None,
                None,
                '0'
            )
        )
        return ParkingAreaEnums.ResultStatusEnums.SUCCESS
